### PR TITLE
feat: add status bar managment

### PR DESCRIPTION
**Pull request by WoLewicki**
Added status bar management on iOS based on the VCs. The user must apply change described in README to `AppDelegate.m` and to the project's `Info.plist` in order for the new props to work.

### CODE REVIEW

**Comment by kmagiera** [comment]
@brentvatne suggested that we don't stick with the core RN's statubar naming convention as it isn't intuitive when accounting for light/dark mode. Instead we could use the following options for the status bar color: 'auto', 'inverted', 'light', or 'dark' (see the new Expo's status bar API: https://docs.expo.io/versions/latest/sdk/status-bar/#statusbarstyle)

File: native-stack/README.md [hunk 1]
```diff
@@ -179,6 +179,31 @@ How the given screen should appear/disappear when pushed or popped at the top of
 
 Defaults to `default`.
 
+### Status bar managment
+
+With `native-stack`, the status bar can be managed by `UIViewController` on iOS. It requires:
+
+1. Enabling (or deleting) `View controller-based status bar appearance` in your Info.plist file (and so disables the option to use React Native's `StatusBar` component).
```
Comment (WoLewicki) [inline native-stack/README.md:186]:
```suggestion
1. Enabling (or deleting) `View controller-based status bar appearance` in your Info.plist file (it disables the option to use React Native's `StatusBar` component). 
```

File: ios/RNSScreen.m [hunk 2]
```diff
@@ -142,6 +148,54 @@ - (void)setStackAnimation:(RNSScreenStackAnimation)stackAnimation
...
+        [self->_controller setNeedsStatusBarAppearanceUpdate];
+      }];
+    }
+  }
+}
+
+- (void)setStatusBarHidden:(BOOL)statusBarHidden
+{
+  if (statusBarHidden != _statusBarHidden) {
+    if (![[[NSBundle mainBundle] objectForInfoDictionaryKey:@"UIViewControllerBasedStatusBarAppearance"] boolValue]) {
```
Comment (kmagiera) [inline ios/RNSScreen.m:186]: Maybe extract this to a separate method? We don't need special warning per parameter. Alternatively it'd be the best to evaluate objectForInfoDictionaryKey call in global scope and store it in a const, mainBundle cannot be changed in runtime so it should be safe to only evaluate it this way

File: ios/RNSScreen.m [hunk 3]
```diff
@@ -333,7 +458,9 @@ - (void)willMoveToParentViewController:(UIViewController *)parent
 - (void)viewWillAppear:(BOOL)animated
 {
   [super viewWillAppear:animated];
-
+  [UIView animateWithDuration:0.5 animations:^{
```
Comment (kmagiera) [inline ios/RNSScreen.m:461]: why 0.5? If this is needed then please add some comment to explain this

File: ios/RNSScreenContainer.m [hunk 4]
```diff
@@ -148,8 +165,6 @@ - (void)updateContainer
           // for screens that were already active we want to mimick the effect UINavigationController
           // has when willMoveToWindow:nil is triggered before the animation starts
           [self prepareDetach:screen];
-          // disable interactions for the duration of transition
```
Comment (kmagiera) [inline ios/RNSScreenContainer.m:151]: these changes seem irrelevant? bad merge?

**Review by kmagiera — COMMENTED** [review]
left a few comments inline

File: ios/RNSScreen.m (hunk 3 above)
Comment (WoLewicki) [reply ios/RNSScreen.m:461]: It is an arbitrary value taken from some SO thread. `Programming iOS 13` book uses a 0.4-second long transition but also doesn't specify why such value exactly, so I have no answer for this.

File: ios/RNSScreen.m [hunk 5]
```diff
@@ -142,6 +149,50 @@ - (void)setStackAnimation:(RNSScreenStackAnimation)stackAnimation
...
+    _controller.modalPresentationCapturesStatusBarAppearance = YES;
+    [UIView animateWithDuration:0.5 animations:^{
+      [self->_controller setNeedsStatusBarAppearanceUpdate];
+    }];
+  }
+}
+
++ (void)checkUIVCBasedStatusBarAppearance
+{
+  if (viewControllerStatusBarDisabled) {
```
Comment (kmagiera) [inline ios/RNSScreen.m:190]: there is no need to define variables outside or mark them as extern. I'd perhaps rename this method to `assertViewControllerBasedStatusBarAppearenceSet` then the implementation should look as follows:


```objc
+ (void)assertViewControllerBasedStatusBarAppearenceSet
{
    static dispatch_once_t once;
    static bool viewControllerBasedAppearence;
    dispatch_once(&once, ^{
viewControllerBasedAppearence = [[[NSBundle mainBundle] objectForInfoDictionaryKey:@"UIViewControllerBasedStatusBarAppearance"] boolValue];
    });
    if (!viewControllerBasedAppearence) {
      RCTLogError(@"........");
    }
}
```

File: ios/RNSScreen.m (hunk 3 above)
Comment (kmagiera) [reply ios/RNSScreen.m:458]: we should have that note included in the code as a comment then (preferably including the link to SO answer)

File: ios/RNSScreenStackHeaderConfig.h [hunk 6]
```diff
@@ -50,7 +60,8 @@ typedef NS_ENUM(NSInteger, RNSScreenStackHeaderSubviewType) {
 
 + (RNSScreenStackHeaderSubviewType)RNSScreenStackHeaderSubviewType:(id)json;
 + (UIBlurEffectStyle)UIBlurEffectStyle:(id)json;
-+ (UISemanticContentAttribute)RNSScreenDirection:(id)json;
++ (UISemanticContentAttribute)UISemanticContentAttribute:(id)json;
```
Comment (WoLewicki) [inline ios/RNSScreenStackHeaderConfig.h:63]: I changed it beacuse the method name was a relict of some of the previous implementations and could be misleading.

File: ios/RNSScreenStack.m [hunk 7]
```diff
@@ -17,6 +18,28 @@ @interface RNSScreenStackView () <UINavigationControllerDelegate, UIAdaptivePres
...
+
+- (UIStatusBarStyle)preferredStatusBarStyle
+{
+  UIViewController *childVC =  [[self childViewControllers] lastObject];
+  return [childVC isKindOfClass:[RNSScreen class]] ? childVC.preferredStatusBarStyle : UIStatusBarStyleDefault;
+}
+
+- (UIStatusBarAnimation)preferredStatusBarUpdateAnimation
+{
+  return [[self childViewControllers] lastObject].preferredStatusBarUpdateAnimation;
```
Comment (kmagiera) [inline ios/RNSScreenStack.m:31]: maybe we should use `topViewController` instead of `childViewControllers lastObject` here and in the previous method?

File: ios/UIViewController+RNScreens.m [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+  return [self reactNativeScreensChildViewControllerForStatusBarHidden];
+}
+
+- (UIStatusBarAnimation)reactNativeScreensPreferredStatusBarUpdateAnimation
+{
+  UIViewController* lastViewController = [[self childViewControllers] lastObject];
+  if ([lastViewController isKindOfClass:[RNScreensNavigationController class]] || [lastViewController isKindOfClass:[RNScreensViewController class]] || [lastViewController isKindOfClass:[RNSScreen class]]) {
+    return lastViewController.preferredStatusBarUpdateAnimation;
+  }
+  return [self reactNativeScreensPreferredStatusBarUpdateAnimation];
```
Comment (kmagiera) [inline ios/UIViewController+RNScreens.m:35]: isn't this an infinite loop?

File: ios/UIViewController+RNScreens.m [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+#import <objc/runtime.h>
+
+
+@implementation UIViewController (RNScreens)
+
+- (UIViewController *)reactNativeScreensChildViewControllerForStatusBarStyle
+{
+  UIViewController* lastViewController = [[self childViewControllers] lastObject];
+  if ([lastViewController isKindOfClass:[RNScreensNavigationController class]] || [lastViewController isKindOfClass:[RNScreensViewController class]] || [lastViewController isKindOfClass:[RNSScreen class]]) {
```
Comment (kmagiera) [inline ios/UIViewController+RNScreens.m:14]: maybe we can define a protocol that all RNScreen navigators conform to in order to avoid all these checks.

File: ios/UIViewController+RNScreens.m [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+    return lastViewController;
```
Comment (kmagiera) [inline ios/UIViewController+RNScreens.m:15]: we have all these swizzled method defined for RNScreensViewController and for RNScreenNavigationController. It seems to me that it'd be better to redirect the call to those implementations instead of having it implemented twice. I understand that what we want to do here is to forward calls to last child controller, but we only want to do it so long `self` is not one of the view controllers defined by screens library (RNScreensVC or RNScreenNavConroller)

**Review by kmagiera — APPROVED** [review]
Left a couple more comments. Accepting to unblock

File: ios/UIViewController+RNScreens.m (hunk 8 above)
Comment (WoLewicki) [reply ios/UIViewController+RNScreens.m:35]: It will call the base method, that is how the swizzling works (based on 
 https://nshipster.com/method-swizzling/#invoking-_cmd)

File: ios/RNSScreen.m [hunk 11]
```diff
@@ -295,6 +297,63 @@ - (instancetype)initWithView:(UIView *)view
...
+  }
+  RNSScreenStackHeaderConfig *config = [self findConfigForScreen];
+  return config && config.statusBarHidden ? config.statusBarHidden : NO;
+}
+
+- (RNSScreenStackHeaderConfig *)findConfigForScreen
+{
+  // if there is no child navigator and the parent is `RNSScreenContainer`, we should fallback to the parent's (that is not `RNSScreenContainer`) option
+  UIViewController *parent = [self parentViewController];
+  while ([parent isKindOfClass:[RNScreensViewController class]] || [parent isKindOfClass:[RNSScreen class]]) {
```
[the remark this answers is not in the document]
Comment (WoLewicki) [reply ios/RNSScreen.m:336]: Would it enter the loop if the `parent` was nil at the beginning? I think this while condition is not that bad since we iterate as long as the parent is one of those two kinds of VC. If we spotted something else, or we got to the top of the hierarchy, loop would stop.

File: ios/RNSScreen.m [hunk 12]
```diff
@@ -295,6 +297,63 @@ - (instancetype)initWithView:(UIView *)view
...
+    if ([parent parentViewController] == nil && [parent isKindOfClass:[RNSScreen class]]) {
+      // we are at the top of hierarchy and the controller is RNSScreen so we are in modal screen
+      break;
+    } else {
+      parent = [parent parentViewController];
+    }
+  }
+  RNSScreenView *screenView = [parent isKindOfClass:[RNScreensNavigationController class]]
```
[the remark this answers is not in the document]
Comment (WoLewicki) [reply ios/RNSScreen.m:344]: If it is not, we will go to the `else` section of the second ternary operator, where we just return the `self.view`, which in the case of `RNSScreen` is always `RNSScreenView`. It probably is good to check if the child of `RNScreensNavigationController` is `RNSScreen` for sure and add comments about the other options.

File: ios/UIViewController+RNScreens.m (hunk 10 above)
Comment (WoLewicki) [reply ios/UIViewController+RNScreens.m:15]: `RNScreenNavigationController` doesn't use the swizzled methods, probably due to `UINavigationController` having own implementations of them, that don't call `super`, so I am not sure if it would help with anything.

Comment (kmagiera) [reply ios/UIViewController+RNScreens.m:14]: How could then it be possible that `lastViewController` is of type `RNSScreen`? We should olways expect it to be of a container type (either RNSSVC or RNSNC)

Comment (WoLewicki) [reply ios/UIViewController+RNScreens.m:14]: These methods can be used in `RNScreensViewController` too iirc.

File: ios/RNSScreen.m (hunk 12 above)
[the remark this answers is not in the document]
Comment (kmagiera) [reply ios/RNSScreen.m]: What if instead of doing this parent traversal we eliminate the need of accessing the config in childViewControllerForStatusBarStyle and others? It looks like it'd be much easier to do the up-to-bottom traversal when we're asked to provide child component as opposed to always returning some child VC and in that child VC doing a reverse traversal to find a parent where config is set

File: ios/RNSScreen.h [hunk 13]
```diff
@@ -51,7 +53,7 @@ typedef NS_ENUM(NSInteger, RNSScreenReplaceAnimation) {
 @property (nonatomic, copy) RCTDirectEventBlock onDismissed;
 @property (nonatomic, copy) RCTDirectEventBlock onWillAppear;
 @property (nonatomic, copy) RCTDirectEventBlock onWillDisappear;
-@property (weak, nonatomic) UIView<RNSScreenContainerDelegate> *reactSuperview;
+@property (weak, nonatomic) UIView <RNSScreenContainerDelegate> *reactSuperview;
```
Comment (kmagiera) [inline ios/RNSScreen.h:56]:
```suggestion
@property (weak, nonatomic) UIView<RNSScreenContainerDelegate> *reactSuperview;
```

**Review by kmagiera — APPROVED** [review]
Thanks for addressing all the previous comments. IMO it is much cleaner than all the previous iterations. Good job 👍

**Comment by jacobp100** [comment]
Thanks for working on this! I'm not on a project that needs it at the moment, but it will be really helpful in the future! 😀

**Comment by a-eid** [comment]
how does this feature work with expo based apps ?

**Comment by brentvatne** [comment]
@a-eid - it's not enabled in Expo Go because we'd have to enable view controller based status bar appearance, but [in the nearish future it should be possible](https://blog.expo.io/expo-managed-workflow-in-2021-d1c9b68aa10)

**Comment by a-eid** [comment]
@brentvatne thank you, I've read about this a few days ago on twitter. 
pretty existing. 

you mentioned that this is not enabled in `Expo Go`, does it work if you build the app ? 
- I've been trying `Statusbar` from RN and `expo-statusbar` and both of them don't work work `native-stack`.

**Comment by brentvatne** [comment]
those will work in native-stack just fine. let me know what issues you encounter with it and include a minimal reproducible example in https://github.com/expo/expo/issues

**Comment by a-eid** [comment]
@brentvatne yes they do, was a mistake on my part. 

thank you

**Comment by brentvatne** [comment]
@a-eid - `RCTStatusBarManager` is a module from upstream react-native. if you don't use React Native's StatusBar API then you won't get the warning: https://github.com/facebook/react-native/blob/1465c8f3874cdee8c325ab4a4916fda0b3e43bdb/React/CoreModules/RCTStatusBarManager.mm#L148-L151

**Comment by Hirbod** [comment]
@brentvatne @a-eid I have the same issue:
I am using an expo-dev-client (0.8.0) and my app is staying black after setting UIViewControllerBasedStatusBarAppearance to true. I searched my whole project and made sure no StatusBar imports are left, also removed expo-status-bar from my project. Sure this isn't an expo issue? Maybe a third party package inside node_modules is causing this...

**Comment by Hirbod** [comment]
@brentvatne I think the issue is coming from here:
https://github.com/expo/expo/blob/master/packages/expo-dev-launcher/bundle/views/LauncherApp.tsx#L42

Thats the only existing StatusBar Import left

**Comment by Hirbod** [comment]
@brentvatne I created an issue for it:
https://github.com/expo/expo/issues/15801
